pydicom/valuerep.py

```diff
@@ -43,6 +43,58 @@
 # (the component separator '=' is handled separately)
 PN_DELIMS = {0x5e}
 
+FIXED_VALUE_VRS = ("AS", "DA")
+
+MAX_VALUE_LEN = {
+    "AE": 16,
+    "AS": 4,
+    "CS": 16,
+    "DA": 8,
+    "DS": 16,
+    "DT": 26,
+    "IS": 12,
+    "LO": 64,
+    "LT": 10240,
+    "PN": 196,
+    "SH": 16,
+    "ST": 1024,
+    "TM": 14,
+    "UI": 64
+}
+
+
+def validate_value(vr: str, value: Any, raise_on_error: bool) -> None:
```

Comment:
I wonder if it'd be better to have VR-specific validators in a `dict` so it could be customised? 

```python
def validate_ae(value: Any) -> Tuple[bool, str]:
    if not 0 <= value <= 16:
        return False, "AE elements must be between 0 and 16 characters long (inclusive)"

    # TODO: Check value is ASCII...

    return True

VALIDATORS = {
    "AE": validate_ae,
    ...,
}

def validate_value(vr, value, raise_on_error):
    validator = VALIDATORS[vr]
    if not validator(value) and raise_on_error:
        raise ValueError(...)
```
Maybe with a provision to allow returning a specific exception message on validation fail

Reply:
Yes, good point. Actually I have been thinking about this, given that there are specific validations for many VRs. I just started with the check that is more less the same for all, so it wasn't really needed, but I'll probably change that.

Reply:
I think it makes a lot of sense.  Could be very much like the [converters](https://github.com/pydicom/pydicom/blob/63b32aaf6cdab92c8dec9bf030d98cd8892ef747/pydicom/values.py#L792) in `values.py`.  There could be generic functions for byte-based max lengths and character-based lengths, with a `maxlength` value passed to them, which would handle most cases.

Reply:
Exactly, this is what I had in mind (except I hadn't been thinking about byte-based and character-based lengths yet). I will change it accordingly, though it may take some time. 
I also first thought to adapt the `config` values as has been discussed, but I abandoned that as I couldn't get it done in an upwards-compatible way that wasn't ugly.